future-apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,569 @@
+        allowedTokens[etherToken] = true;
+        allowedTokensArray.push(etherToken);
+        if (address(etherToken) != denominationToken) {
+            allowedTokens[denominationToken] = true;
+            allowedTokensArray.push(denominationToken);
+        }
+    }
+
+    /**
+     * @dev Set the Denomination exchange rate for a token. Uses decimals from token
+     * @notice Sets the Denomination exchange rate for a token
+     * @param token The token address
+     * @param denominationExchangeRate The exchange rate
+     */
+    function setExchangeRate(address token, uint256 denominationExchangeRate) external auth(ORACLE_ROLE) {
+        // Denomination Token is a special one, so we can not allow its exchange rate to be changed
+        if (token != denominationToken) {
+            exchangeRates[token] = denominationExchangeRate;
+            LogSetExchangeRate(token, denominationExchangeRate);
+        }
+    }
+
+    /**
+     * @dev Add tokens to the allowed set
+     * @notice Add tokens to the allowed set
+     * @param _allowedTokens Array of new tokens allowed for payment
+     */
+    function addAllowedTokens(address[] _allowedTokens) public auth(EMPLOYER_ROLE) {
+        for (uint i = 0; i < _allowedTokens.length; i++) {
+            if (!allowedTokens[_allowedTokens[i]]) {
+                allowedTokens[_allowedTokens[i]] = true;
+                allowedTokensArray.push(_allowedTokens[i]);
+            }
+        }
+    }
+
+    /*
+     * TODO: removeFromAllowedTokens. It wouldn't be trivial, as employees
+     * should modifiy their allocation. They should be notified and their
+     * last allocation date should be reset.
+     */
+
+    /**
+     * @dev Add employee to Payroll
+     * @notice Add employee to Payroll. See addEmployeeWithNameAndStartDate
+     * @param accountAddress Employee's address to receive Payroll
+     * @param initialYearlyDenominationSalary Employee's salary
+     */
+    function addEmployee(
+        address accountAddress,
+        uint256 initialYearlyDenominationSalary
+    )
+        public
+        auth(EMPLOYER_ROLE)
+    {
+        _addEmployee(
+            accountAddress,
+            initialYearlyDenominationSalary,
+            "",
+            getTimestamp()
+        );
+    }
+
+    /**
+     * @dev Add employee to Payroll
+     * @notice Add employee to Payroll. See addEmployeeWithNameAndStartDate
+     * @param accountAddress Employee's address to receive Payroll
+     * @param initialYearlyDenominationSalary Employee's salary
+     * @param name Employee's name
+     */
+    function addEmployeeWithName(
+        address accountAddress,
+        uint256 initialYearlyDenominationSalary,
+        string name
+    )
+        public
+        auth(EMPLOYER_ROLE)
+    {
+        _addEmployee(
+            accountAddress,
+            initialYearlyDenominationSalary,
+            name,
+            getTimestamp()
+        );
+    }
+
```

Comment:
I just thought that we could make this contract a forwarder, that forwards scripts to all employees in an org. We need to add the finance app to the blacklist so they can't steal all the moneyz though (See token manager forwarding!)